### PR TITLE
use constant time mem_cmp to compare sensitive data

### PR DESCRIPTION
**Pull request by vixentael**
when checking private keys and inside secure separator

### CODE REVIEW

File: src/themis/secure_comparator.c [hunk 1]
```diff
@@ -169,7 +171,7 @@ static themis_status_t ed_verify(uint8_t pos, const ge_p3 *point, const uint8_t
 
 	sc_reduce(k);
 
-	if (memcmp(k, signature, ED25519_GE_LENGTH))
+	if (cst_time_memcmp(k, signature, ED25519_GE_LENGTH))
```
Comment (secumod) [inline src/themis/secure_comparator.c]: Not sure this is needed: this is signature verification and signature is public here.

File: src/themis/secure_comparator.c [hunk 2]
```diff
@@ -302,7 +304,7 @@ static themis_status_t ed_dbl_base_verify(uint8_t pos, const ge_p3 *base1, const
...
```
Comment (secumod) [inline src/themis/secure_comparator.c]: ditto

File: src/themis/secure_comparator.c [hunk 3]
```diff
@@ -438,7 +440,7 @@ static themis_status_t ed_point_verify(uint8_t pos, const ge_p3 *base2, const ge
...
```
Comment (secumod) [inline src/themis/secure_comparator.c]: ditto

File: src/soter/boringssl/soter_ec_key.c [hunk 4]
```diff
@@ -373,7 +374,7 @@ soter_status_t soter_ec_priv_key_to_engine_specific(const soter_container_hdr_t
 	}
 
 	/* Validate tag */
-	if (memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
+	if (cst_time_memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/soter/boringssl/soter_ec_key.c]: This is public info as well (formatting prefix)

File: src/soter/boringssl/soter_rsa_key.c [hunk 5]
```diff
@@ -443,7 +444,7 @@ soter_status_t soter_rsa_priv_key_to_engine_specific(const soter_container_hdr_t
 	}
 
 	/* Validate tag */
-	if (memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
+	if (cst_time_memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/soter/boringssl/soter_rsa_key.c]: This is public

File: src/soter/openssl/soter_ec_key.c [hunk 6]
```diff
@@ -372,7 +373,7 @@ soter_status_t soter_ec_priv_key_to_engine_specific(const soter_container_hdr_t
 	}
 
 	/* Validate tag */
-	if (memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
+	if (cst_time_memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/soter/openssl/soter_ec_key.c]: This is public

File: src/soter/openssl/soter_rsa_key.c [hunk 7]
```diff
@@ -443,7 +444,7 @@ soter_status_t soter_rsa_priv_key_to_engine_specific(const soter_container_hdr_t
 	}
 
 	/* Validate tag */
-	if (memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
+	if (cst_time_memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/soter/openssl/soter_rsa_key.c]: This is public

File: src/themis/secure_message_wrapper.c [hunk 8]
```diff
@@ -39,10 +41,10 @@ soter_sign_alg_t get_alg_id(const uint8_t* key, size_t key_length)
   if (key_length < sizeof(soter_container_hdr_t) && key_length < ((const soter_container_hdr_t*)key)->size){
     return (soter_sign_alg_t)(-1);
   }
-  if (memcmp(((const soter_container_hdr_t*)key)->tag,EC_PRIV_KEY_PREF,3)==0 || memcmp(((const soter_container_hdr_t*)key)->tag,EC_PUB_KEY_PREF,3)==0){
+  if (cst_time_memcmp(((const soter_container_hdr_t*)key)->tag,EC_PRIV_KEY_PREF,3)==0 || memcmp(((const soter_container_hdr_t*)key)->tag,EC_PUB_KEY_PREF,3)==0){
```
Comment (secumod) [inline src/themis/secure_message_wrapper.c]: this is public

File: src/themis/secure_message_wrapper.c [hunk 9]
```diff
@@ -39,10 +41,10 @@ soter_sign_alg_t get_alg_id(const uint8_t* key, size_t key_length)
...
       return SOTER_SIGN_ecdsa_none_pkcs8;
   }
-  if (memcmp(((const soter_container_hdr_t*)key)->tag,RSA_PRIV_KEY_PREF,3)==0 || memcmp(((const soter_container_hdr_t*)key)->tag,RSA_PUB_KEY_PREF,3)==0){
+  if (cst_time_memcmp(((const soter_container_hdr_t*)key)->tag,RSA_PRIV_KEY_PREF,3)==0 || memcmp(((const soter_container_hdr_t*)key)->tag,RSA_PUB_KEY_PREF,3)==0){
```
Comment (secumod) [inline src/themis/secure_message_wrapper.c]: public

File: src/themis/secure_session_utils.c [hunk 10]
```diff
@@ -37,12 +38,12 @@ soter_sign_alg_t get_key_sign_type(const void *sign_key, size_t sign_key_length)
 			return (soter_sign_alg_t)0xffffffff;
 		}
 
-		if (!memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
+		if (!cst_time_memcmp(key->tag, EC_PRIV_KEY_PREF, strlen(EC_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/themis/secure_session_utils.c]: public

File: src/themis/secure_session_utils.c [hunk 11]
```diff
@@ -37,12 +38,12 @@ soter_sign_alg_t get_key_sign_type(const void *sign_key, size_t sign_key_length)
...
 		{
 			return SOTER_SIGN_ecdsa_none_pkcs8;
 		}
 
-		if (!memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
+		if (!cst_time_memcmp(key->tag, RSA_PRIV_KEY_PREF, strlen(RSA_PRIV_KEY_PREF)))
```
Comment (secumod) [inline src/themis/secure_session_utils.c]: public

File: src/themis/secure_session_utils.c [hunk 12]
```diff
@@ -190,7 +191,7 @@ themis_status_t verify_mac(const void *key, size_t key_length, const soter_kdf_c
 		return THEMIS_INVALID_PARAMETER;
 	}
 
-	if (memcmp(mac, computed_mac, mac_length))
+	if (cst_time_memcmp(mac, computed_mac, mac_length))
```
Comment (vixentael) [inline src/themis/secure_session_utils.c]: @secumod is this the only valid check here? or mac is public too?

**Comment by vixentael** [comment]
@secumod can you please check my latest comment, and close issue if you feel that adding constant time mem cmp is not relevant in these cases?

https://github.com/cossacklabs/themis/pull/245#discussion_r147983115

File: src/themis/secure_session_utils.c (hunk 12 above)
Comment (secumod) [reply src/themis/secure_session_utils.c]: Hm... Seems it is public as well

**Comment by secumod** [comment]
Yes. Taking another look - seems it is not needed here, since all the parameters are public.

**Comment by vixentael** [comment]
thank you :)

**Pull request closed without merging**